vllm/compilation/qk_norm_rope_fusion.py

```diff
@@ -0,0 +1,278 @@
+from vllm.config import VllmConfig, get_layers_from_vllm_config
+from vllm.logger import init_logger
+from vllm.platforms import current_platform
+
+from .fusion import empty_bf16, empty_i64
+from .inductor_pass import enable_fake_mode
+from .vllm_inductor_pass import VllmInductorPass, VllmPatternMatcherPass
+
+logger = init_logger(__name__)
+
+
+# Ops used in the pattern (assume kernels are built and available)
+RMS_OP = torch.ops._C.rms_norm.default
+ROPE_OPS: list[torch._ops.OpOverload] = [
+    torch.ops._C.rotary_embedding.default,
+    # torch.ops.vllm.flashinfer_rotary_embedding.default,
+]
+FUSED_QK_ROPE_OP = torch.ops._C.fused_qk_norm_rope.default
+SPLIT_SIZES_OP = torch.ops.aten.split_with_sizes.default
+RESHAPE_OP = torch.ops.aten.reshape.default
+EMPTY_LIKE_OP = torch.ops.aten.empty_like.default
+VIEW_OP = torch.ops.aten.view.default
+CONTIGUOUS_OP = torch.ops.aten.contiguous.default
+
+
+class QkNormRopePattern:
+    """
+    Match the unfused sequence in attention blocks and replace with the fused op.
+
+    Unfused (conceptually):
+      q, k, v = split(qkv, [qsz, kvsz, kvsz], -1)
+      qh = reshape(q, [-1, num_heads, head_dim])
+      kh = reshape(k, [-1, num_kv_heads, head_dim])
+      qn = rms_norm(qh, q_weight, eps)
+      kn = rms_norm(kh, k_weight, eps)
+      qf = reshape(qn, [-1, num_heads * head_dim])
+      kf = reshape(kn, [-1, num_kv_heads * head_dim])
+      qf, kf = rotary_embedding(positions, qf, kf, head_dim, cos_sin_cache, is_neox)
+      return qf, kf, v
+
+    Fused replacement:
+      fused_qk_norm_rope(qkv, num_heads, num_kv_heads, num_kv_heads, head_dim,
+                         eps, q_weight, k_weight, cos_sin_cache, is_neox,
+                         positions.view(-1))
+      return split(qkv, [qsz, kvsz, kvsz], -1)
+    """
+
+    def __init__(
+        self,
+        layer: Attention,
+        eps: float,
+        rope_op: torch._ops.OpOverload,
+        is_neox: bool,
+    ) -> None:
+        self.layer = layer
+        self.num_heads = layer.num_heads
+        self.num_kv_heads = layer.num_kv_heads
+        self.head_dim = layer.head_size
+        self.q_size = self.num_heads * self.head_dim
+        self.kv_size = self.num_kv_heads * self.head_dim
+        self.eps = eps
+        self.rope_op = rope_op
+        self.is_neox = is_neox
+
+    def register(self, pm_pass: PatternMatcherPass):
+        def pattern(
+            qkv: torch.Tensor,
+            positions: torch.Tensor,
+            q_weight: torch.Tensor,
+            k_weight: torch.Tensor,
+            cos_sin_cache: torch.Tensor,
+        ):
+            # split qkv -> q,k,v
+            # q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
+            split_tuple = SPLIT_SIZES_OP(
+                qkv, [self.q_size, self.kv_size, self.kv_size], -1
+            )
+            q = operator.getitem(split_tuple, 0)
+            k = operator.getitem(split_tuple, 1)
+            v = operator.getitem(split_tuple, 2)
```

Comment:
I think that this should work, pattern tracing is very close to forward code tracing:
```suggestion
            # split qkv -> q,k,v
            q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
```